File: YEAR 2/SEM 1/Advanced Programming Methods/LAB 4/src/Model/Statement/AssignmentStatement.java
package Model.Statement;

import Model.Expression.ExpressionException;
import Model.Expression.IExpression;
import Model.State.ProgramState;
import Util.Dictionary.MyIDictionary;

public class AssignmentStatement implements IStatement {
    private String variable;
    private IExpression expression;

    public AssignmentStatement(String variable, IExpression expression){
        this.variable = variable;
        this.expression = expression;
    }

    @Override
    public ProgramState execute(ProgramState programState) throws ExpressionException {
        MyIDictionary<String,Integer> symbolTable = programState.getSymbolTable();

        int result = this.expression.evaluate(symbolTable);

        // update the value of the variable
        symbolTable.put(this.variable,result);

        return programState;
    }

    @Override
    public String toString() {
        return "("+ this.variable + "=" + this.expression+")";
    }
}

File: YEAR 2/SEM 1/Advanced Programming Methods/ToyLanguage/ToyLanguage/Model/Statement/AssignmentStatement.cs
using ToyLanguage.Model.Expression;
using ToyLanguage.Model.State;
using ToyLanguage.Util.Dictionary;

namespace ToyLanguage.Model.Statement
{
    class AssignmentStatement : IStatement
    {
        private readonly string id;
        private readonly IExpression expression;

        public AssignmentStatement(string id,IExpression expression)
        {
            this.id = id;
            this.expression = expression;
        }

        public ProgramState Execute(ProgramState programState)
        {
            MyIDictionary<string, int> symbolTable = programState.SymbolTable;

            int value = expression.Evaluate(symbolTable);

            if(symbolTable.ContainsKey(id))
            {
                symbolTable.Update(id, value);
            }
            else
            {
                symbolTable.Add(id, value);
            }

            return programState;
        }

        public override string ToString()
        {
            return "AssignmentStatement(" + id + "," + expression + ")";
        }
    }
}
